lte/gateway/c/oai/tasks/sgw/pgw_handlers.c

```diff
@@ -536,68 +563,67 @@ uint32_t pgw_handle_nw_initiated_bearer_deactv_req(
   OAILOG_FUNC_IN(LOG_PGW_APP);
   MessageDef *message_p = NULL;
   uint32_t i = 0;
-  uint32_t j = 0;
+  uint32_t itrn = 0;
   hash_table_ts_t *hashtblP = NULL;
   uint32_t num_elements = 0;
   s_plus_p_gw_eps_bearer_context_information_t *spgw_ctxt_p = NULL;
   hash_node_t *node = NULL;
   itti_s5_nw_init_deactv_bearer_request_t *itti_s5_deactv_ded_bearer_req = NULL;
-  bool found = false;
+  bool is_lbi_found = false;
+  bool send_reject = false;
+  bool is_imsi_found = false;
+  bool is_ebi_found = false;
+  ebi_t ebi_to_be_established[BEARERS_PER_UE] = {0};
+  uint32_t no_of_bearers_to_be_est = 0;
+  uint32_t no_of_bearers_rej = 0;
+  ebi_t invalid_ebi[BEARERS_PER_UE] = {0};
+  teid_t s11_mme_teid = 0;
 
   OAILOG_INFO(LOG_PGW_APP, "Received nw_initiated_deactv_bearer_req from NW\n");
   print_bearer_ids_helper(bearer_req_p->ebi, bearer_req_p->no_of_bearers);
-  message_p = itti_alloc_new_message(
-    TASK_SPGW_APP, S5_NW_INITIATED_DEACTIVATE_BEARER_REQ);
-  if (message_p == NULL) {
-    OAILOG_ERROR(
-      LOG_PGW_APP,
-      "itti_alloc_new_message failed for nw_initiated_deactv_bearer_req\n");
-    OAILOG_FUNC_RETURN(LOG_PGW_APP, RETURNerror);
-  }
 
-  itti_s5_deactv_ded_bearer_req =
-    &message_p->ittiMsg.s5_nw_init_deactv_bearer_request;
-  //Send ITTI message to SGW
-  memset(
-    itti_s5_deactv_ded_bearer_req,
-    0,
-    sizeof(itti_s5_nw_init_deactv_bearer_request_t));
-  itti_s5_deactv_ded_bearer_req->delete_default_bearer = false;
-  itti_s5_deactv_ded_bearer_req->no_of_bearers = bearer_req_p->no_of_bearers;
-  memcpy(
-    &itti_s5_deactv_ded_bearer_req->ebi,
-    bearer_req_p->ebi,
-    (sizeof(ebi_t)) * bearer_req_p->no_of_bearers);
+
   hashtblP = spgw_state->sgw_state.s11_bearer_context_information;
   if (hashtblP == NULL) {
     OAILOG_ERROR(
       LOG_PGW_APP, "hashtblP is NULL for nw_initiated_deactv_bearer_req\n");
     OAILOG_FUNC_RETURN(LOG_PGW_APP, RETURNerror);
   }
 
-  //Check if EBI recvd == LBI to know if default bearer has to be deactivated
+  //Check if valid LBI and EBI recvd
   while ((num_elements < hashtblP->num_elements) && (i < hashtblP->size) &&
-         (!found)) {
+         (!is_ebi_found)) {
```

Comment:
How about if is_imsi_found was true but not is_ebi_found. Should this stop?

Reply:
No it should not. It will search the entire hashlist to find the EBI entry

Reply:
No, if we already found the IMSI and it did not have the EBI we are looking for there is no value in checking other IMSIs.
See comment below